cmd/cli/version.go

```diff
@@ -1,32 +1,146 @@
+			var err error
+
+			cliVersionInfo := version.GetInfo()
+			versionInfo.cliVersionInfo = &cliVersionInfo
+
+			if versionCmd.remote {
+				err = versionCmd.setKubeClientset()
+				if err == nil {
+					versionCmd.namespace = settings.Namespace()
+					versionCmd.remoteVersion = &remoteVersion{}
+					versionInfo.remoteVersionInfo, err = versionCmd.getMeshVersion()
+				}
+			}
+
+			versionCmd.outputVersionInfo(versionInfo)
+			if err != nil {
+				fmt.Fprintf(versionCmd.out, "Failed to get mesh version: %v", err)
+			}
 		},
 	}
+
+	f := cmd.Flags()
+	f.BoolVarP(&versionCmd.remote, "remote", "r", true, "If false, shows only the OSM CLI version")
+
 	return cmd
 }
+
+func (v *versionCmd) setKubeClientset() error {
+	config, err := settings.RESTClientGetter().ToRESTConfig()
+	if err != nil {
+		return errors.Errorf("Error fetching kubeconfig: %s", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Errorf("Could not access Kubernetes cluster, check kubeconfig: %s", err)
+	}
+	v.clientset = clientset
+	return nil
+}
+
+func (v *versionCmd) getMeshVersion() (*remoteVersionInfo, error) {
+	var version *version.Info
+	var versionInfo *remoteVersionInfo
+
+	controllerPods, err := getControllerPods(v.clientset, v.namespace)
+	if err != nil {
+		return nil, err
+	}
+	if len(controllerPods.Items) == 0 {
+		return nil, errors.Errorf("No mesh found in namespace [%s]", v.namespace)
+	}
+
+	controllerPod := controllerPods.Items[0]
+	version, err = v.remoteVersion.proxyGetMeshVersion(controllerPod.Name, v.namespace, v.clientset)
+	if err != nil {
+		return nil, err
+	}
+	versionInfo = &remoteVersionInfo{
+		meshName: controllerPod.Labels[constants.OSMAppInstanceLabelKey],
+		version:  version,
+	}
+	return versionInfo, nil
+}
+
+func (r *remoteVersion) proxyGetMeshVersion(pod string, namespace string, clientset kubernetes.Interface) (*version.Info, error) {
+	resp, err := clientset.CoreV1().Pods(namespace).ProxyGet("", pod, strconv.Itoa(constants.OSMHTTPServerPort), httpserverconstants.VersionPath, nil).DoRaw(context.TODO())
+	if err != nil {
+		return nil, errors.Errorf("Error retrieving mesh version from pod %s in namespace %s: %s", pod, namespace, err)
+	}
+	if len(resp) == 0 {
+		return nil, errors.Errorf("Error retrieving mesh version from pod %s in namespace %s", pod, namespace)
+	}
+
+	versionInfo := &version.Info{}
+	err = json.Unmarshal(resp, versionInfo)
+	if err != nil {
+		return nil, errors.Errorf("Error unmarshalling retrieved mesh version from pod %s in namespace %s: %s", pod, namespace, err)
+	}
+
+	return versionInfo, nil
+}
+
+func (v *versionCmd) outputVersionInfo(versionInfo versionInfo) {
+	fmt.Fprintf(v.out, "CLI Version: %s\n", fmt.Sprintf("%#v", *versionInfo.cliVersionInfo))
+	if v.remote && versionInfo.remoteVersionInfo != nil {
+		fmt.Fprintf(v.out, "Mesh [%s] Version: %s\n", versionInfo.remoteVersionInfo.meshName, fmt.Sprintf("%#v", *versionInfo.remoteVersionInfo.version))
```

Comment:
"Mesh" could be replaced with  "Server", "Remote", or "Control Plane". Would appreciate any feedback!